cxplat/inc/cxplat_memory.h

```diff
@@ -52,60 +45,49 @@ __drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_writes_maybenull_(size) void*
  * @param[in] initialize False to return "uninitialized" memory.
  * @returns Pointer to memory block allocated, or null on failure.
  */
-__drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_writes_maybenull_(size) void* cxplat_allocate_with_tag(
+__drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_writes_maybenull_(size) void* cxplat_allocate(
     _In_ cxplat_pool_type_t pool_type, size_t size, uint32_t tag, bool initialize);
 
 /**
  * @brief Reallocate memory.
  * @param[in] memory Allocation to be reallocated.
  * @param[in] old_size Old size of memory to reallocate.
  * @param[in] new_size New size of memory to reallocate.
- * @returns Pointer to memory block allocated, or null on failure.
- */
-__drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_writes_maybenull_(new_size) void* cxplat_reallocate(
-    _In_ _Post_invalid_ void* memory, size_t old_size, size_t new_size);
-
-/**
- * @brief Reallocate memory with tag.
- * @param[in] memory Allocation to be reallocated.
- * @param[in] old_size Old size of memory to reallocate.
- * @param[in] new_size New size of memory to reallocate.
  * @param[in] tag Pool tag to use.
  * @returns Pointer to memory block allocated, or null on failure.
  */
-__drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_writes_maybenull_(new_size) void* cxplat_reallocate_with_tag(
+__drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_writes_maybenull_(new_size) void* cxplat_reallocate(
     _In_ _Post_invalid_ void* memory, size_t old_size, size_t new_size, uint32_t tag);
 
+// Pseudo-tag used with free API if the caller does not know the pool tag.
+// This is used for example by the usersim library to implement APIs such as
+// WdfObjectDelete() and ExFreePool().
+#define CXPLAT_TAG_ANY 0
```

Comment:
I don't quite understand this requirement, and I'm not a fan that this make the free call require that extra branch every time to check for this special case and then call a different function. There is no way around this? Can we not have a special case function just for this then `cxplat_free_any_tag()`?

Reply:
We can, but again the compiler optimizer should optimize it out.  But if you prefer it to be a separate API, I can do that.

Reply:
Done

Reply:
Thanks.